src/main/scala/ftanml/util/JsonReader.scala

```
@@ -0,0 +1,49 @@
+package ftanml.util
+
+import java.io.Reader
+import java.util.Stack
+
+class JsonReader(base: Reader) extends Reader {
+	var depthMapsArray = new Stack[Int] //0: Map, 1: Array (tried to do it with an enum but couldn't get that working
+	var escaped = false
+	var inString = false
+	
+	override def read(): Int = {
+		val char = base.read()
+		char match {
+			case '"' => if(!escaped) inString = !inString; '"'
+			case '\\' => if(inString) escaped = !escaped; '\\'
+			case _ if(inString) => char
+			case '{' => depthMapsArray.push(0); '<'
+			case '}' => assert(depthMapsArray.pop == 0, "Expected ], found }"); '>'
```

I'd rather throw an exception here, because this can happen on invalid input. Asserts won't throw exception in production environments.